### PR TITLE
✨ Is922/add rabbitMQ to resource usage tracker

### PR DESCRIPTION
**Pull request by matusdrobuliak66**
## What do these changes do?
- ✨ add rabbitMQ to the resource usage tracker
- ♻️ put the previously intended logic of scraping Prometheus to separate module "Prometheus containers" for now. And comment it out in the application core. 

In upcoming PRs, we will create/connect to the rabbitMQ queue:
image

## Related issue/s

- relates https://github.com/ITISFoundation/osparc-issues/issues/922
- prerequisite https://github.com/ITISFoundation/osparc-issues/issues/1051

## How to test
```
cd services/resource-usage-tracker
make install-dev
make test-dev
```

### CODE REVIEW

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/modules/rabbitmq.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    if not app.state.rabbitmq_client:
+        raise ConfigurationError(
+            msg="RabbitMQ client is not available. Please check the configuration."
+        )
+    return cast(RabbitMQClient, app.state.rabbitmq_client)
+
+
+async def post_message(app: FastAPI, message: RabbitMessageBase) -> None:
+    with log_catch(logger, reraise=False), contextlib.suppress(ConfigurationError):
+        # NOTE: if rabbitmq was not initialized the error does not need to flood the logs
```
Comment (sanderegg) [inline services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/modules/rabbitmq.py:49]: does that really make sense? are you allowing it to start without rabbitmq?
and do you actually post_messages at the moment?

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/cli.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
 from settings_library.utils_cli import create_settings_command
 from simcore_service_resource_usage_tracker.core.errors import ConfigurationError
 from simcore_service_resource_usage_tracker.modules.prometheus import create_client
-from simcore_service_resource_usage_tracker.resource_tracker_cli_placeholder import (
+from simcore_service_resource_usage_tracker.modules.prometheus_containers.cli_placeholder import (
```
Comment (sanderegg) [inline services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/cli.py:12]: relative imports (on all 3 lines here)

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py [hunk 3]
```diff
@@ -2,47 +2,52 @@
...
 
 _logger = logging.getLogger(__name__)
 
 
+async def _subscribe_to_rabbitmq(app) -> None:
+    with log_context(_logger, logging.INFO, msg="Subscribing to rabbitmq channel"):
+        rabbit_client: RabbitMQClient = get_rabbitmq_client(app)
+
+        # TODO: subscribe to rabbitmq channel
+        print("rabbit_client", rabbit_client)
```
Comment (sanderegg) [inline services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py:19]: remove print

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py [hunk 4]
```diff
@@ -2,47 +2,52 @@
...
+
+
+async def _unsubscribe_from_rabbitmq(app) -> None:
+    with log_context(
+        _logger, logging.INFO, msg="Unsubscribing from rabbitmq channels"
+    ), log_catch(_logger, reraise=False):
+        rabbit_client: RabbitMQClient = get_rabbitmq_client(app)
+
+        # TODO: unsubscribe from rabbitmq channel
+        print("rabbit_client", rabbit_client)
```
Comment (sanderegg) [inline services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py:31]: remove print

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py (hunk 3 above)
Comment (sanderegg) [reply services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py:19]: 2. you can already connect to the correct exchanges:
_RabbitResourceTrackingBaseMessage.get_channel_name()

File: services/resource-usage-tracker/tests/unit/modules/prometheus_containers/test_resource_tracker.py [hunk 5]
```diff
@@ -32,16 +32,20 @@ def app_environment(
...
     mocked_task = mocker.patch(
-        "simcore_service_resource_usage_tracker.resource_tracker.collect_container_resource_usage_task",
+        "simcore_service_resource_usage_tracker.modules.prometheus_containers.plugin.collect_container_resource_usage_task",
         autospec=True,
     )
     return mocked_task
 
 
+@pytest.mark.skip(
+    reason="This test is currently not needed, as setup_background_task is commented out in application.py"
```
Comment (sanderegg) [inline services/resource-usage-tracker/tests/unit/modules/prometheus_containers/test_resource_tracker.py:42]: is there any point in keeping it in the code? or do you plan to add it again? if not just remove dead code while you still know it is to be removed.

**Review by sanderegg — COMMENTED** [review]
Looking good.
2 things:
- you can already connect to the rabbitmq exchanges and test it
- please do not keep dead code if you plan on totally removing it or not using it in the near future.

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/modules/rabbitmq.py (hunk 1 above)
Comment (matusdrobuliak66) [reply services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/modules/rabbitmq.py:49]: you are right, i don't need it at the moment, removing :+1:

File: services/resource-usage-tracker/tests/unit/modules/prometheus_containers/test_resource_tracker.py (hunk 5 above)
Comment (matusdrobuliak66) [reply services/resource-usage-tracker/tests/unit/modules/prometheus_containers/test_resource_tracker.py:42]: I would still keep it for a while, I have a feeling we will utilize also Prometheus in the near future.

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py (hunk 3 above)
Comment (matusdrobuliak66) [reply services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py:19]: I have one question here, we can discuss it in person.

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/modules/rabbitmq.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+def setup(app: FastAPI) -> None:
+    async def on_startup() -> None:
+        app.state.rabbitmq_client = None
+        settings: RabbitSettings | None = (
+            app.state.settings.RESOURCE_USAGE_TRACKER_RABBITMQ
+        )
+        if not settings:
+            logger.warning("Rabbit MQ client is de-activated in the settings")
+            return
```
[the remark this answers is not in the document]
Comment (matusdrobuliak66) [reply services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/modules/rabbitmq.py:22]: yes you are right, i have change it, thanks!

File: services/resource-usage-tracker/tests/unit/modules/prometheus_containers/test_resource_tracker.py (hunk 5 above)
Comment (matusdrobuliak66) [reply services/resource-usage-tracker/tests/unit/modules/prometheus_containers/test_resource_tracker.py:42]: As discussed in person, for now, we will leave it and we will see as it evolves in the near future. Then we will refactor and remove not used code.

File: packages/service-library/src/servicelib/rabbitmq.py [hunk 7]
```diff
@@ -184,8 +184,8 @@ async def subscribe(
 
         assert self._channel_pool  # nosec
         async with self._channel_pool.acquire() as channel:
-            _DEFAULT_PREFETCH_VALUE = 10  # this value is set to the default for now
-            await channel.set_qos(_DEFAULT_PREFETCH_VALUE)
+            _PREFETCH_VALUE = 1 if exclusive_queue is False else 10
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/rabbitmq.py:187]: Please set the 10 as an external constant, such as _DEFAULT_PREFETCH_VALUE. we might want to adjust that in the future

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/core/application.py [hunk 8]
```diff
@@ -50,8 +51,13 @@ def create_app(settings: ApplicationSettings) -> FastAPI:
     if settings.RESOURCE_USAGE_TRACKER_POSTGRES:
         setup_db(app)
     setup_redis(app)
+    setup_rabbitmq(app)
 
-    setup_background_task(app)
+    # NOTE: This task currently runs in the background and collects data from prometheus for all
+    # jupyter smash services. This is currently not needed.
+    # setup_background_task(app)
```
Comment (sanderegg) [inline services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/core/application.py:58]: just remove it. git is here for history.

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py [hunk 9]
```diff
@@ -1,48 +1,62 @@
...
+
+    _logger.info("Received message")
+    return True
+
+
+async def _subscribe_to_rabbitmq(app) -> None:
+    with log_context(_logger, logging.INFO, msg="Subscribing to rabbitmq channel"):
+        rabbit_client: RabbitMQClient = get_rabbitmq_client(app)
+        await rabbit_client.subscribe(
+            _RabbitResourceTrackingBaseMessage.get_channel_name(),
```
Comment (sanderegg) [inline services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py:27]: change that private class to become public. is pylint not complaining?

**Review by sanderegg — APPROVED** [review]
Good for me, plus please check the 3 code smells from sonar cloud.

File: packages/service-library/src/servicelib/rabbitmq.py (hunk 7 above)
Comment (matusdrobuliak66) [reply packages/service-library/src/servicelib/rabbitmq.py:187]: done :+1:

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/core/application.py (hunk 8 above)
Comment (matusdrobuliak66) [reply services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/core/application.py:58]: done :+1:

File: services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py (hunk 9 above)
Comment (matusdrobuliak66) [reply services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/resource_tracker.py:27]: done :+1:

**Review by GitHK — APPROVED** [review]

**Pull request merged**